### PR TITLE
Allow specifying nodeSelector, affinity and tolerations

### PR DESCRIPTION
**Pull request by pkliczewski**
As part of placement rules we need to extend vm import config and update controller deployment.

Signed-off-by: Piotr Kliczewski <piotr.kliczewski@gmail.com>

### CODE REVIEW

File: pkg/apis/v2v/v1alpha1/vmimportconfig_types.go [hunk 1]
```diff
@@ -25,6 +25,33 @@ import (
 // VMImportConfigSpec defines the desired state of VMImportConfig
 type VMImportConfigSpec struct {
 	ImagePullPolicy corev1.PullPolicy `json:"imagePullPolicy,omitempty"`
+
+	// Rules on which nodes controller pod(s) will be scheduled
+	Infra NodePlacement `json:"infra,omitempty"`
```
Comment (jakub-dzon) [inline pkg/apis/v2v/v1alpha1/vmimportconfig_types.go:31]: Should we touch v1alpha any more? We have newer v1beta.

**Review by jakub-dzon — COMMENTED** [review]
Looks good - just one question.

File: pkg/apis/v2v/v1alpha1/vmimportconfig_types.go (hunk 1 above)
Comment (pkliczewski) [reply pkg/apis/v2v/v1alpha1/vmimportconfig_types.go:31]: This is what I saw in CDI so I did the same. In the past we had to fix `v1alpha` to have successful HCO deployment.

Comment (jakub-dzon) [reply pkg/apis/v2v/v1alpha1/vmimportconfig_types.go:31]: OK

**Review by jakub-dzon — APPROVED** [review]

**Review by jakub-dzon — CHANGES_REQUESTED** [review]
Please fix the CI build.

File: pkg/apis/v2v/v1alpha1/vmimportconfig_types.go (hunk 1 above)
Comment (masayag) [inline pkg/apis/v2v/v1alpha1/vmimportconfig_types.go:31]: should this field be mandatory ?

File: pkg/apis/v2v/v1beta1/vmimportconfig_types.go [hunk 2]
```diff
@@ -25,6 +25,33 @@ import (
 // VMImportConfigSpec defines the desired state of VMImportConfig
 type VMImportConfigSpec struct {
 	ImagePullPolicy corev1.PullPolicy `json:"imagePullPolicy,omitempty"`
+
+	// Rules on which nodes controller pod(s) will be scheduled
+	Infra NodePlacement `json:"infra,omitempty"`
```
Comment (masayag) [inline pkg/apis/v2v/v1beta1/vmimportconfig_types.go:31]: should this field be mandatory ?
Also, what 'Infra' stands for in this context ? perhaps just 'Placement' or 'NodePlacement' which describes the nature of this field as is ?

**Review by jakub-dzon — APPROVED** [review]

File: pkg/apis/v2v/v1alpha1/vmimportconfig_types.go (hunk 1 above)
Comment (pkliczewski) [reply pkg/apis/v2v/v1alpha1/vmimportconfig_types.go:31]: Based on HCO design my understanding is that if it is not there we rely on scheduler to place the controller pod on any node. In this context it is OK to have it optional.

File: pkg/apis/v2v/v1beta1/vmimportconfig_types.go (hunk 2 above)
Comment (pkliczewski) [reply pkg/apis/v2v/v1beta1/vmimportconfig_types.go:31]: As @jakub-dzon is working on extracting common code from both CDI and vm-import I used the same name of the type as CDI to make it easy to extract.

File: pkg/apis/v2v/v1alpha1/vmimportconfig_types.go (hunk 1 above)
Comment (pkliczewski) [reply pkg/apis/v2v/v1alpha1/vmimportconfig_types.go:31]: Added  // +optional  for both

File: pkg/operator/controller/controller.go [hunk 3]
```diff
@@ -775,6 +776,9 @@ func (r *ReconcileVMImportConfig) getOperatorArgs(cr *v2vv1.VMImportConfig) *Ope
 		if cr.Spec.ImagePullPolicy != "" {
 			result.PullPolicy = string(cr.Spec.ImagePullPolicy)
 		}
+		if &cr.Spec.Infra != nil {
```
Comment (masayag) [inline pkg/operator/controller/controller.go:779]: If you define Infra 
`NodePlacement `json:"infra,omitempty"``
as
`Infra *NodePlacement `json:"infra,omitempty"``

you could just check 
```
if cr.Spec.Infra != nil {
...
```

without need to check the address with &. This seems to complete the API notation of having an optional element.
Is there any specific reason not to define the NodePlacement on the VMImportConfig as *NodePlacement ?

Comment (pkliczewski) [reply pkg/operator/controller/controller.go:779]: Wtih you suggestion compiler complains:
> pkg/operator/controller/controller.go:779:20: cannot convert nil to type "github.com/kubevirt/vm-import-operator/pkg/apis/v2v/v1beta1".NodePlacement

Comment (pkliczewski) [reply pkg/operator/controller/controller.go:779]: The only reason is consistency with CDI.

Comment (masayag) [reply pkg/operator/controller/controller.go:779]: I'm referring to the guidelines on https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#optional-vs-required

Comment (masayag) [reply pkg/operator/controller/controller.go:779]: > The only reason is consistency with CDI.

okay  - so we can stick to it.

**Review by masayag — APPROVED** [review]

**Pull request merged**